main.go

```diff
@@ -81,80 +114,101 @@ func run(ctx context.Context, castaiclient castai.Client, baseLogger *logrus.Log
+		clusterID := ""
+		if cfg.Static != nil {
+			clusterID = cfg.Static.ClusterID
+		}
 
-	if clusterID == "" {
-		reg, err := provider.RegisterCluster(ctx, castaiclient)
-		if err != nil {
-			return fmt.Errorf("registering cluster: %w", err)
+		if clusterID == "" {
+			reg, err := provider.RegisterCluster(ctx, castaiclient)
+			if err != nil {
+				return fmt.Errorf("registering cluster: %w", err)
+			}
+			clusterID = reg.ClusterID
+			clusterIDChanged(clusterID)
+			log.Infof("cluster registered: %v, clusterID: %s", reg, clusterID)
+		} else {
+			clusterIDChanged(clusterID)
+			log.Infof("clusterID: %s provided by env variable", clusterID)
 		}
-		clusterID = reg.ClusterID
-		log.Infof("cluster registered: %v, clusterID: %s", reg, clusterID)
-	} else {
-		log.Infof("clusterID: %s provided by env variable", clusterID)
-	}
-	castailog.SetupLogExporter(baseLogger, localLog, castaiclient, castailog.Config{
-		ClusterID:   clusterID,
-		SendTimeout: LogExporterSendTimeout,
-	})
 
-	log.Data["cluster_id"] = clusterID
+		w := &controller.Worker{
+			Fn: controller.RunController(log, clientset, castaiclient, provider, clusterID, cfg, agentVersion, ctrlHealthz),
+		}
+		defer w.Stop(log)
 
-	exitCh := make(chan error)
+		if err := w.Start(ctx); err != nil {
+			return fmt.Errorf("controller loop error: %w", err)
+		}
 
-	if cfg.PprofPort != 0 {
-		addr := fmt.Sprintf(":%d", cfg.PprofPort)
-		pprofSrv := &http.Server{Addr: addr, Handler: http.DefaultServeMux}
-		defer func() {
-			if err := pprofSrv.Close(); err != nil {
-				log.Errorf("closing pprof server: %v", err)
+		return nil
+	}
+
+	replicas.Run(ctx, log, cfg.LeaderElection, clientset, leaderWatchDog, func(ctx context.Context) {
+		exitCh <- leaderFunc(ctx)
+	})
+	return nil
+}
+
+// if any errors are observed on exitCh, context cancel is called, and all errors in the channel are logged
+func watchExitErrors(ctx context.Context, log *logrus.Entry, exitCh chan error, ctxCancel func()) {
+	select {
+	case err := <-exitCh:
+		if err != nil && !errors.Is(err, context.Canceled) {
+			log.Errorf("agent stopped with an error: %v", err)
+		}
+		ctxCancel()
+		for err := range exitCh {
+			if err != nil && !errors.Is(err, context.Canceled) {
+				log.Errorf("additional shutdown errors: %v", err)
 			}
-		}()
+		}
+	case <-ctx.Done():
+		return
+	}
+}
 
-		go func() {
-			log.Infof("starting pprof server on %s", addr)
-			exitCh <- fmt.Errorf("pprof server: %w", pprofSrv.ListenAndServe())
-		}()
+func runPProf(cfg config.Config, log *logrus.Entry, exitCh chan error) (closeFunc func()) {
+	addr := portToServerAddr(cfg.PprofPort)
+	pprofSrv := &http.Server{Addr: addr, Handler: http.DefaultServeMux}
+	closeFn := func() {
+		if err := pprofSrv.Close(); err != nil {
+			log.Errorf("closing pprof server: %v", err)
+		}
 	}
 
-	ctrlHealthz := controller.NewHealthzProvider(cfg)
+	go func() {
+		log.Infof("starting pprof server on %s", addr)
+		exitCh <- fmt.Errorf("pprof server: %w", pprofSrv.ListenAndServe())
```

Comment:
Because we're now logging all exit errors, won't this always produce an error on shutdown? When `Close()` is called on an `http.Server`, the `ListenAndServe` returns an `ErrServerClosed` error which should be ignored.

Reply:
hm.. haven't observed this with manual testing, will check